File: Chain of Responsibility/Golang/Caching.go
package main

import (
	"fmt"
)

type Caching struct {
	nextHandler Handler
	data        map[string]*Response
}

func (caching *Caching) setNextHandler(handler Handler) {
	caching.nextHandler = handler
}

func (caching *Caching) execute(request *Request) *Response {
	if caching.data == nil {
		caching.data = make(map[string]*Response)
	}

	if _, ok := caching.data[request.Route]; ok {
		fmt.Println("Cached")

		return caching.data[request.Route]
	} else {
		fmt.Println("Uncached")

		temp := caching.nextHandler.execute(request)

		caching.data[request.Route] = temp

		return temp
	}
}

File: Chain of Responsibility/Dart/Caching.dart
import 'Handler.dart';
import 'Response.dart';
import 'Request.dart';

class Caching extends Handler {
  Map<String?, Response?> _cache = {};

  @override
  Response? execute(Request? request) {
    if (_cache[request?.get_route] != null) {
      print("Cached");

      return _cache[request?.get_route];
    } else {
      print("Uncached");

      Response? temp = next_handler?.execute(request);
      _cache[request?.get_route] = temp;

      return temp;
    }
  }
}
